sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -609,7 +609,13 @@ object DataSource extends Logging {
 
   /** Given a provider name, look up the data source class definition. */
   def lookupDataSource(provider: String, conf: SQLConf): Class[_] = {
-    val provider1 = backwardCompatibilityMap.getOrElse(provider, provider) match {
+    val customBackwardCompatibilityMap =
+      conf.getAllConfs
+        .filter(_._1.startsWith("spark.sql.datasource.map"))
+        .map{ case (k, v) => (k.replaceFirst("^spark.sql.datasource.map.", ""), v) }
+    val compatibilityMap = backwardCompatibilityMap ++ customBackwardCompatibilityMap
```

Comment:
so at this point we are leaving com.databricks.spark.avro -> internal avro as the default and users have to set it back to com.databricks.spark.avro or do they set it empty?   Although if set to empty I think it will return empty below which will cause an issue.

We should have a test case for empty and perhaps have a check for it below in that case.

What about documentation? Is there a jira for documenting all avro stuff?  If we do leave it as default we definitely want a release note with change in behavior.

Reply:
answering one of my own questions found avro docs here: https://github.com/apache/spark/pull/22121

Reply:
If this is merged, we can remove the internal Avro mapping code and put that into documents before branch cut.

> so at this point we are leaving com.databricks.spark.avro -> internal avro as the default and users have to set it back to com.databricks.spark.avro or do they set it empty?

This will be better than another option like the following `avro` specific option.
```scala
val ENABLE_AVRO_BACKWARD_COMPATIBILITY = 
  buildConf("spark.sql.avro.backwardCompatibility")
```

Reply:
I have the same concern as @tgravescs . It seems tricky to unset the default mapping.

For example, if by default we map `com.databricks.spark.avro` to  internal avro, then to unset it we have to set
`spark.sql.datasource.map.com.databricks.spark.avro -> com.databricks.spark.avro` .

Currently we only have to deal with Avro and CSV, so I think it is ok to have one single straightforward configuration like https://github.com/apache/spark/pull/22133 proposed.